src/lamp_py/performance_manager/l1_rt_trips.py

```diff
@@ -326,12 +329,145 @@ def update_start_times(db_manager: DatabaseManager) -> None:
+    )
+
+    red_events = (
+        sa.select(
+            VehicleEvents.pm_trip_id,
+            VehicleEvents.stop_id,
+        )
+        .select_from(VehicleEvents)
+        .join(
+            distinct_red_trips,
+            sa.and_(
+                distinct_red_trips.c.service_date == VehicleEvents.service_date,
+                distinct_red_trips.c.pm_trip_id == VehicleEvents.pm_trip_id,
+            ),
+        )
+    )
+
+    process_logger = ProcessLogger("l1_trips.update_branch_trunk_route_id")
+    process_logger.log_start()
+
+    trips_df = db_manager.select_as_dataframe(distinct_trips)
+    red_events_df = db_manager.select_as_dataframe(red_events)
+
+    def get_red_branch(pm_trip_id: int) -> Optional[str]:
+        ashmont_stop_ids = {
+            "70087",
+            "70088",
+            "70089",
+            "70090",
+            "70091",
+            "70092",
+            "70093",
+            "70094",
+            "70085",
+            "70086",
+        }
+        braintree_stop_ids = {
+            "70097",
+            "70098",
+            "70099",
+            "70100",
+            "70101",
+            "70102",
+            "70103",
+            "70104",
+            "70105",
+            "70095",
+            "70096",
+        }
+        trip_stop_ids = set(
+            red_events_df[red_events_df["pm_trip_id"] == pm_trip_id]["stop_id"]
+        )
+        if trip_stop_ids & ashmont_stop_ids:
+            return "Red-A"
+        if trip_stop_ids & braintree_stop_ids:
+            return "Red-B"
+        return None
+
+    def map_trunk_route_id(route_id: str) -> str:
+        if str(route_id).startswith("Green"):
+            return "Green"
+        return route_id
+
+    def apply_branch_route_id(record: pandas.Series) -> Optional[str]:
+        if str(record["route_id"]).startswith("Green"):
+            return record["route_id"]
+        if str(record["route_id"]) == "Red":
+            return get_red_branch(record["pm_trip_id"])
+        return None
+
+    trips_df["trunk_route_id"] = trips_df["route_id"].map(map_trunk_route_id)
+    trips_df["branch_route_id"] = trips_df.apply(apply_branch_route_id, axis=1)
+
+    trips_df = trips_df[
+        [
+            "pm_trip_id",
+            "branch_route_id",
+            "trunk_route_id",
+        ]
+    ].rename(
+        columns={
+            "pm_trip_id": "b_pm_trip_id",
+            "branch_route_id": "b_branch_route_id",
+            "trunk_route_id": "b_trunk_route_id",
+        }
+    )
+
+    start_times_update_query = (
+        sa.update(VehicleTrips.__table__)
+        .where(VehicleTrips.pm_trip_id == bindparam("b_pm_trip_id"))
+        .values(
+            branch_route_id=bindparam("b_branch_route_id"),
+            trunk_route_id=bindparam("b_trunk_route_id"),
+        )
+    )
+
+    db_manager.execute_with_data(
+        start_times_update_query,
+        trips_df,
+        disable_trip_tigger=True,
+    )
+    process_logger.log_complete()
```

Comment:
i generally find these types of DB manipulations easier to follow. not sure what that says about me. i've often wondered about doing things this way vs some of our more complicated db computes.